aten/src/ATen/native/nestedtensor/NestedTensorImpl.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+#include <ATen/ATen.h>
+#include <ATen/MemoryOverlap.h>
+#include <c10/core/MemoryFormat.h>
+#include <c10/util/Metaprogramming.h>
+
+namespace at {
+namespace native {
+
+bool is_nested_tensor_impl(const at::Tensor& tensor) {
+  return tensor.unsafeGetTensorImpl()->key_set().has(
+      c10::DispatchKey::NestedTensor);
+}
+
+struct NestedTensorImpl : public c10::TensorImpl {
+  explicit NestedTensorImpl(at::Tensor buffer, at::Tensor nested_size_tensor);
+
+  int64_t dim() const override {
+    return _nested_size_tensor.dim() > 0 ? 1 + _nested_size_tensor.size(1) : 1;
```

Comment:
instead of overriding this, can we simply just allocate size_ field with the appropriate number of entries (even if those entries contain garbage?)

Reply:
We can. Let's do that once we find this operation is slow in the context of a set of operations and it shows up in a profile. For now I want to aim for correctness.

Reply:
This is about correctness though? Scott has devirtualized many of the size related methods on TensorImpl (a development I wholeheartedly support), and so custom overrides like this will work differently inside fbcode. If we can dispense with the override entirely we can also guarantee identical behavior in OSS and fbcode.

BTW, I've kind of gotten to the point where I think nested tensors should have an actual size, and simply MAX_INT for any ragged dimensions. I think this will work pretty well, since you cannot actually exceed MAX_INT entries in the ragged representation. MAX_INT is also stupidly large and you will get an error if you try to use it for another allocation.

Reply:
Sure we can add that, but it would be great if calling .size() on NestedTensorImpl were to actually throw, because I've been bitten in the past by calling into composite operators or other unsupported operations unknowingly and getting very weird errors (for example matmul will call size). Any NestedTensor kernel that wants to use size() unconditionally is doing something wrong (note that there is an interpretation for size(i) if i is regular). I've worked around this in pytorch/nestedtensor by wrapping call devirtualized method calls into e.g. ```get_size```, ```get_dtype``` etc.